README.md

```diff
@@ -1,31 +1,784 @@
+
+`StatusCodeResponseVerifier` is preimplemented `ResponseVerifier` that compares status code of response with codes from `init`. If match is not found then it returns `.invalidStatusCode`, nil otherwise.
+
+### InputEncoding
+
+`InputEncoding` is a marker protocol which says how the data should be encoded to `Request` and how decoded from `Response`. Each `Endpoint` has default encoding based on the method it uses (`GET` encodes data to the url, `POST` to the request body atc.) but this can be changed by passing different encoding in its `init` (see [Endpoint](#endpoint)).
+
+#### InputEncodingWithEncoder
+
+```Swift
+protocol InputEncodingWithEncoder: InputEncoding {
+    
+    func encode(input: SupportedType, to request: inout Request)
+}
+```
+
+`InputEncodingWithEncoder` is an extension of `InputEncoding`. Normally `DataEncoder` takes care of encoding the input data based on `InputEncoding` (see [DataEncoder](#dataencoder)). With this protocol the `encode` method is called to do the job instead.
+
+#### StandardInputEncoding
+
+```Swift
+enum StandardInputEncoding: InputEncoding {
+    
+    case queryString
+    case httpBody
+}
+```
+
+`StandardInputEncoding` represents two encodings that all `DataEncoder` understand. 
+
+`.queryString` encodes the input data into the request url (`http://url.xxx?param1=value`). As shown in examples in [Quick overview](#quick-overview) this may also be done manually. `.queryString` has a disadvantage, that the input data must be a dictionary. 
+
+`.httpBody` simply sends the data in the request body. Concrete encoding depends on `DataEncoder` implementation (may be JSON, XML, etc.).
+
+### RequestPerformer
+
+```Swift
+protocol RequestPerformer {
+    
+    var dataEncoder: DataEncoder { get }
+    
+    func perform(request: Request, callback: @escaping (Response<Data>) -> Void) -> Cancellable
+}
+```
+
+`RequestPerformer` duty is to communicate with the server (perform requests). This is done in `perform` method that takes `Request` and calls `callback` with `Response` (this call may not happen immediately), it returns `Cancellable` (way to cancel the request).  
+
+`RequestPerformer` also specifies `DataEncoder`, that will be used to encode and decode the data.
+
+#### DataEncoder
+
+```Swift
+protocol DataEncoder {
+    
+    func encodeToQueryString(input: SupportedType, to request: inout Request)
+    
+    func encodeToHttpBody(input: SupportedType, to request: inout Request)
+    
+    func encodeCustom(input: SupportedType, to request: inout Request, inputEncoding: InputEncoding)
+    
+    func decode(response: Response<Data>) -> Response<SupportedType>
+}
+```
+
+`DataEncoder` is called by `Fetcher` to encode the input data into `Request` (`encode???`) and to decode data from `Response` (`decode`). 
+
+`encodeToQueryString` and `encodeToHttpBody` are used if `StandardInputEncoding` is used (method selected based on value). 
```

Comment:
"Based on value (optional), `StandardInputEncoding` specifies which of `encodeToQueryString` or `encodeToHttpBody` is used."